lib/mongoid/persistence_context.rb

```diff
@@ -232,7 +231,39 @@ def clear(object, cluster = nil, original_context = nil)
           end
         end
       ensure
-        Thread.current["[mongoid][#{object.object_id}]:context"] = original_context
+        store_context(object, original_context)
+      end
+
+      private
+
+      # Get the persistence context for a given object from the thread local
+      #   storage.
+      #
+      # @param [ Object ] object Object to get the persistance context for.
+      #
+      # @return [ Mongoid::PersistenceContext | nil ] The persistence context
+      #   for the object if previously stored, otherwise nil.
+      #
+      # @api private
+      def get_context(object)
+        Thread.current["mongoid:persistence_context"] ||= {}
+        Thread.current["mongoid:persistence_context"][object.object_id]
+      end
+
+      # Store persistence context for a given object in the thread local
+      #   storage.
+      #
+      # @param [ Object ] object Object to store the persistance context for.
+      # @param [ Mongoid::PersistenceContext ] context Context to store
+      #
+      # @api private
+      def store_context(object, context)
+        Thread.current["mongoid:persistence_context"] ||= {}
```

Comment:
This line can be moved into the `else` path and line 263 changed to

```
          Thread.current["mongoid:persistence_context"]&.delete(object.object_id)
``

Reply:
Done